# Chuẩn hóa dữ liệu Google Play Store

## Mục lục
1. Chọn dataset
2. Dataset có gì
3. Đánh giá vấn đề
4. Đưa ra giải pháp
5. Kiểm chứng

In [1]:
import pandas as pd
import numpy as np

# Đọc dữ liệu
df = pd.read_csv("googleplaystore.csv")
print(f"Kích thước: {df.shape[0]} dòng x {df.shape[1]} cột")
df.head()

Kích thước: 10841 dòng x 13 cột


,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


## 1. Chọn Dataset

- **Tên:** Google Play Store Apps
- **Nguồn:** Kaggle (Lavanya Gupta)
- **Mô tả:** Chứa thông tin các ứng dụng trên Google Play Store
- **Số dòng:** ~10,841
- **Số cột:** 13

## 2. Dataset có gì?

In [2]:
print("=== Tên cột và kiểu dữ liệu ===")
print(df.dtypes)
print("\n=== Thông tin tổng quan ===")
df.info()

=== Tên cột và kiểu dữ liệu ===
App                   str
Category              str
Rating            float64
Reviews               str
Size                  str
Installs              str
Type                  str
Price                 str
Content Rating        str
Genres                str
Last Updated          str
Current Ver           str
Android Ver           str
dtype: object

=== Thông tin tổng quan ===
<class 'pandas.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  str    
 1   Category        10841 non-null  str    
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  str    
 4   Size            10841 non-null  str    
 5   Installs        10841 non-null  str    
 6   Type            10840 non-null  str    
 7   Price           10841 non-null  str    
 8   Content Rating  10840 non-null  st

In [3]:
print("=== 5 dòng đầu ===")
display(df.head())
print("\n=== 5 dòng cuối ===")
display(df.tail())

=== 5 dòng đầu ===


,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up



=== 5 dòng cuối ===


,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
10836,Sya9a Maroc - FR,FAMILY,4.5,38,53M,"5,000+",Free,0,Everyone,Education,"July 25, 2017",1.48,4.1 and up
10837,Fr. Mike Schmitz Audio Teachings,FAMILY,5.0,4,3.6M,100+,Free,0,Everyone,Education,"July 6, 2018",1.0,4.1 and up
10838,Parkinson Exercices FR,MEDICAL,NaN,3,9.5M,"1,000+",Free,0,Everyone,Medical,"January 20, 2017",1.0,2.2 and up
10839,The SCP Foundation DB fr nn5n,BOOKS_AND_REFERENCE,4.5,114,Varies with device,"1,000+",Free,0,Mature 17+,Books & Reference,"January 19, 2015",Varies with device,Varies with device
10840,iHoroscope - 2018 Daily Horoscope & Astrology,LIFESTYLE,4.5,398307,19M,"10,000,000+",Free,0,Everyone,Lifestyle,"July 25, 2018",Varies with device,Varies with device


In [4]:
print("=== Giá trị thiếu ===")
missing = df.isnull().sum()
missing_pct = (df.isnull().sum() / len(df)) * 100
missing_df = pd.DataFrame({'Số thiếu': missing, 'Tỷ lệ %': missing_pct.round(2)})
display(missing_df[missing_df['Số thiếu'] > 0])

=== Giá trị thiếu ===


,Số thiếu,Tỷ lệ %
Rating,1474,13.60
Type,1,0.01
Content Rating,1,0.01
Current Ver,8,0.07
Android Ver,3,0.03


In [5]:
print("=== Dòng trùng lặp ===")
print(f"Số dòng trùng: {df.duplicated().sum()}")
print(f"\nSố App trùng tên: {df['App'].duplicated().sum()}")

=== Dòng trùng lặp ===
Số dòng trùng: 483

Số App trùng tên: 1181


## 3. Đánh giá vấn đề

In [6]:
print("=== Vấn đề 1: Cột Reviews chứa giá trị không phải số ===")
reviews_invalid = df[pd.to_numeric(df['Reviews'], errors='coerce').isna()]
print(f"Số dòng lỗi: {len(reviews_invalid)}")
if len(reviews_invalid) > 0:
    display(reviews_invalid.head())

print("\n=== Vấn đề 2: Cột Size có đơn vị khác nhau (M, k) ===")
print(df['Size'].apply(lambda x: str(x)[-1] if pd.notna(x) else 'NaN').value_counts())

print("\n=== Vấn đề 3: Cột Installs có ký tự đặc biệt (+, ,) ===")
print(df['Installs'].unique()[:10])

print("\n=== Vấn đề 4: Cột Price có ký tự $ ===")
print(df['Price'].apply(lambda x: str(x)[0] if pd.notna(x) else 'NaN').value_counts())

print("\n=== Vấn đề 5: Cột Last Updated là chuỗi ===")
print(df['Last Updated'].head(3))

print("\n=== Vấn đề 6: Rating ngoài phạm vi (0-5) ===")
print(df['Rating'].describe())
invalid_rating = df[(df['Rating'] < 0) | (df['Rating'] > 5)]
print(f"Số dòng Rating ngoài [0, 5]: {len(invalid_rating)}")
if len(invalid_rating) > 0:
    display(invalid_rating)

=== Vấn đề 1: Cột Reviews chứa giá trị không phải số ===
Số dòng lỗi: 1


,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
10472,Life Made WI-Fi Touchscreen Photo Frame,1.9,19.0,3.0M,"1,000+",Free,0,Everyone,NaN,"February 11, 2018",1.0.19,4.0 and up,NaN



=== Vấn đề 2: Cột Size có đơn vị khác nhau (M, k) ===
Size
M    8829
e    1695
k     316
+       1
Name: count, dtype: int64

=== Vấn đề 3: Cột Installs có ký tự đặc biệt (+, ,) ===
<StringArray>
[     '10,000+',     '500,000+',   '5,000,000+',  '50,000,000+',
     '100,000+',      '50,000+',   '1,000,000+',  '10,000,000+',
       '5,000+', '100,000,000+']
Length: 10, dtype: str

=== Vấn đề 4: Cột Price có ký tự $ ===
Price
0    10040
$      800
E        1
Name: count, dtype: int64

=== Vấn đề 5: Cột Last Updated là chuỗi ===
0     January 7, 2018
1    January 15, 2018
2      August 1, 2018
Name: Last Updated, dtype: str

=== Vấn đề 6: Rating ngoài phạm vi (0-5) ===
count    9367.000000
mean        4.193338
std         0.537431
min         1.000000
25%         4.000000
50%         4.300000
75%         4.500000
max        19.000000
Name: Rating, dtype: float64
Số dòng Rating ngoài [0, 5]: 1


,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
10472,Life Made WI-Fi Touchscreen Photo Frame,1.9,19.0,3.0M,"1,000+",Free,0,Everyone,NaN,"February 11, 2018",1.0.19,4.0 and up,NaN


In [7]:
print("=== Vấn đề 7: Genres có dấu ; (multi-category) ===")
multi_genre = df[df['Genres'].str.contains(';', na=False)]
print(f"Số dòng có nhiều Genres: {len(multi_genre)}")

print("\n=== Vấn đề 8: Current Ver và Android Ver có giá trị 'Varies with device' ===")
print(f"Current Ver 'Varies with device': {(df['Current Ver'] == 'Varies with device').sum()}")
print(f"Android Ver 'Varies with device': {(df['Android Ver'] == 'Varies with device').sum()}")

=== Vấn đề 7: Genres có dấu ; (multi-category) ===
Số dòng có nhiều Genres: 498

=== Vấn đề 8: Current Ver và Android Ver có giá trị 'Varies with device' ===
Current Ver 'Varies with device': 1459
Android Ver 'Varies with device': 1362


## 4. Giải pháp

In [8]:
# Tạo bản sao để xử lý
df_clean = df.copy()
print(f"Trước clean: {df_clean.shape}")

Trước clean: (10841, 13)


In [9]:
# 4.1: Xóa dòng trùng lặp hoàn toàn
df_clean = df_clean.drop_duplicates()
print(f"Sau xóa trùng: {df_clean.shape}")

Sau xóa trùng: (10358, 13)


In [10]:
# 4.2: Sửa cột Reviews - tìm và xóa dòng không phải số
reviews_invalid = pd.to_numeric(df_clean['Reviews'], errors='coerce').isna()
print(f"Số dòng Reviews lỗi: {reviews_invalid.sum()}")
df_clean = df_clean[~reviews_invalid].copy()
df_clean['Reviews'] = df_clean['Reviews'].astype(int)

Số dòng Reviews lỗi: 1


In [11]:
# 4.3: Xóa dòng có Rating ngoài [0, 5]
df_clean = df_clean[(df_clean['Rating'] >= 0) & (df_clean['Rating'] <= 5)]
print(f"Sau lọc Rating: {df_clean.shape}")

Sau lọc Rating: (8892, 13)


In [12]:
# 4.4: Chuẩn hóa cột Size (M -> số, k -> /1024)
def convert_size(val):
    if pd.isna(val):
        return np.nan
    val = str(val).strip()
    if val == 'Varies with device':
        return np.nan
    if val.endswith('M'):
        return float(val[:-1])
    if val.endswith('k'):
        return float(val[:-1]) / 1024
    return np.nan

df_clean['Size_MB'] = df_clean['Size'].apply(convert_size)
print(f"Size hợp lệ: {df_clean['Size_MB'].notna().sum()}/{len(df_clean)}")

Size hợp lệ: 7424/8892


In [13]:
# 4.5: Chuẩn hóa Installs (bỏ +, ,)
df_clean['Installs_Num'] = (df_clean['Installs']
    .str.replace('+', '', regex=False)
    .str.replace(',', '', regex=False)
    .astype(int))
print(df_clean['Installs_Num'].describe())

count    8.892000e+03
mean     1.648965e+07
std      8.637600e+07
min      1.000000e+00
25%      1.000000e+04
50%      5.000000e+05
75%      5.000000e+06
max      1.000000e+09
Name: Installs_Num, dtype: float64


In [14]:
# 4.6: Chuẩn hóa Price (bỏ $)
df_clean['Price_Num'] = (df_clean['Price']
    .str.replace('$', '', regex=False)
    .astype(float))
print(df_clean['Price_Num'].describe())

count    8892.000000
mean        0.963155
std        16.189341
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max       400.000000
Name: Price_Num, dtype: float64


In [15]:
# 4.7: Chuyển Last Updated sang datetime
df_clean['Last_Updated_Date'] = pd.to_datetime(df_clean['Last Updated'], errors='coerce')
print(df_clean['Last_Updated_Date'].head())

0   2018-01-07
1   2018-01-15
2   2018-08-01
3   2018-06-08
4   2018-06-20
Name: Last_Updated_Date, dtype: datetime64[us]


In [16]:
# 4.8: Điền giá trị thiếu cho Rating bằng trung bình theo Category
df_clean['Rating'] = df_clean.groupby('Category')['Rating'].transform(
    lambda x: x.fillna(x.mean())
)
print(f"Rating còn thiếu: {df_clean['Rating'].isna().sum()}")

Rating còn thiếu: 0


In [17]:
# 4.9: Tách Genres chính (lấy phần đầu trước dấu ;)
df_clean['Main_Genre'] = df_clean['Genres'].str.split(';').str[0]
print(df_clean['Main_Genre'].value_counts().head())

Main_Genre
Tools            734
Entertainment    538
Education        528
Action           364
Productivity     334
Name: count, dtype: int64


In [18]:
# 4.10: Chuẩn hóa chuỗi (xóa khoảng trắng thừa)
for col in df_clean.select_dtypes(include=['object', 'str']).columns:
    df_clean[col] = df_clean[col].astype(str).str.strip()

print(f"\nKích thước sau clean: {df_clean.shape}")


Kích thước sau clean: (8892, 18)


## 5. Kiểm chứng

In [19]:
print("=== KIỂM CHỨNG SAU CLEAN ===")

# 5.1: Không còn dòng trùng
print(f"\n1. Dòng trùng: {df_clean.duplicated().sum()} (yêu cầu: 0)")

# 5.2: Reviews là số nguyên
print(f"2. Reviews kiểu int: {df_clean['Reviews'].dtype}")

# 5.3: Rating trong [0, 5]
print(f"3. Rating range: [{df_clean['Rating'].min():.2f}, {df_clean['Rating'].max():.2f}]")

# 5.4: Installs_Num là số nguyên
print(f"4. Installs_Num kiểu: {df_clean['Installs_Num'].dtype}")
print(f"   Installs_Num min/max: {df_clean['Installs_Num'].min()}/{df_clean['Installs_Num'].max()}")

# 5.5: Price_Num là số
print(f"5. Price_Num kiểu: {df_clean['Price_Num'].dtype}")
print(f"   Có giá trị âm: {(df_clean['Price_Num'] < 0).sum()}")

# 5.6: Last_Updated_Date là datetime
print(f"6. Last_Updated_Date kiểu: {df_clean['Last_Updated_Date'].dtype}")

# 5.7: Size_MB là số
print(f"7. Size_MB kiểu: {df_clean['Size_MB'].dtype}")
print(f"   Size_MB range: [{df_clean['Size_MB'].min():.2f}, {df_clean['Size_MB'].max():.2f}]")

=== KIỂM CHỨNG SAU CLEAN ===

1. Dòng trùng: 0 (yêu cầu: 0)
2. Reviews kiểu int: int64
3. Rating range: [1.00, 5.00]
4. Installs_Num kiểu: int64
   Installs_Num min/max: 1/1000000000
5. Price_Num kiểu: float64
   Có giá trị âm: 0
6. Last_Updated_Date kiểu: datetime64[us]
7. Size_MB kiểu: float64
   Size_MB range: [0.01, 100.00]


In [20]:
print("=== Giá trị thiếu còn lại ===")
missing_final = df_clean.isnull().sum()
print(missing_final[missing_final > 0])

=== Giá trị thiếu còn lại ===
Current Ver       4
Android Ver       2
Size_MB        1468
dtype: int64


In [21]:
# 5.8: So sánh trước/sau
print("=== SO SÁNH TRƯỚC/SAU ===")
print(f"{'Chỉ số':<25}{'Trước':>15}{'Sau':>15}")
print("-" * 55)
print(f"{'Số dòng':<25}{len(df):>15}{len(df_clean):>15}")
print(f"{'Dòng trùng':<25}{df.duplicated().sum():>15}{df_clean.duplicated().sum():>15}")
print(f"{'Reviews lỗi':<25}{pd.to_numeric(df['Reviews'], errors='coerce').isna().sum():>15}{0:>15}")
print(f"{'Rating ngoài [0,5]':<25}{len(df[(df['Rating']<0)|(df['Rating']>5)]):>15}{len(df_clean[(df_clean['Rating']<0)|(df_clean['Rating']>5)]):>15}")

=== SO SÁNH TRƯỚC/SAU ===
Chỉ số                             Trước            Sau
-------------------------------------------------------
Số dòng                            10841           8892
Dòng trùng                           483              0
Reviews lỗi                            1              0
Rating ngoài [0,5]                     1              0


In [22]:
# Lưu file đã clean
df_clean.to_csv("googleplaystore_cleaned.csv", index=False)
print("✅ Đã lưu: googleplaystore_cleaned.csv")
display(df_clean.head())

✅ Đã lưu: googleplaystore_cleaned.csv


,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,Size_MB,Installs_Num,Price_Num,Last_Updated_Date,Main_Genre
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,19.0,10000,0.0,2018-01-07,Art & Design
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,14.0,500000,0.0,2018-01-15,Art & Design
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,8.7,5000000,0.0,2018-08-01,Art & Design
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,25.0,50000000,0.0,2018-06-08,Art & Design
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,2.8,100000,0.0,2018-06-20,Art & Design
